bigquery/query.go

```diff
@@ -324,11 +326,97 @@ func (q *Query) newJob() (*bq.Job, error) {
+// is used in place of the jobs.insert path as this path does not expose a job
+// object.
+func (q *Query) Read(ctx context.Context) (it *RowIterator, err error) {
+	ctx = trace.StartSpan(ctx, "cloud.google.com/go/bigquery.Query.Run")
+	defer func() { trace.EndSpan(ctx, err) }()
+	queryRequest, err := q.probeFastPath()
+	if err != nil {
+		// Any error means we fallback to the older mechanism.
+		job, err := q.Run(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return job.Read(ctx)
+	}
+	// we have a config, run on fastPath.
+	resp, err := q.client.runQuery(ctx, queryRequest)
 	if err != nil {
 		return nil, err
 	}
-	return job.Read(ctx)
+	// construct a minimal job for backing the row iterator.
+	minimalJob := &Job{
+		c:         q.client,
+		jobID:     resp.JobReference.JobId,
+		location:  resp.JobReference.Location,
+		projectID: resp.JobReference.ProjectId,
+	}
+	if resp.JobComplete {
+		rowSource := &rowSource{
+			j: minimalJob,
+			// RowIterator can precache results from the iterator to save a lookup.
+			cachedRows:      resp.Rows,
+			cachedSchema:    resp.Schema,
+			cachedNextToken: resp.PageToken,
+		}
+		return newRowIterator(ctx, rowSource, fetchPage), nil
+	}
+	// We're on the fastPath, but we need to poll because the job is incomplete.
+	// Fallback to job-based Read().
+	return minimalJob.Read(ctx)
+}
+
+// probeFastPath is used to attempt configuring a jobs.Query request based on a
+// user's Query configuration.  If all the options set on the job are supported on the
+// faster query path, this method returns a QueryRequest suitable for execution.
+func (q *Query) probeFastPath() (*bq.QueryRequest, error) {
+	// This is a denylist of settings which prevent us from composing an equivalent
+	// bq.QueryRequest due to differences between configuration parameters accepted
+	// by jobs.insert vs jobs.query.
+	if q.QueryConfig.Dst != nil ||
+		q.QueryConfig.TableDefinitions != nil ||
+		q.QueryConfig.CreateDisposition != "" ||
+		q.QueryConfig.WriteDisposition != "" ||
+		!(q.QueryConfig.Priority == "" || q.QueryConfig.Priority == InteractivePriority) ||
+		q.QueryConfig.UseLegacySQL ||
+		q.QueryConfig.MaxBillingTier != 0 ||
+		q.QueryConfig.TimePartitioning != nil ||
+		q.QueryConfig.RangePartitioning != nil ||
+		q.QueryConfig.Clustering != nil ||
+		q.QueryConfig.DestinationEncryptionConfig != nil ||
+		q.QueryConfig.SchemaUpdateOptions != nil ||
+		// User has defined the jobID generation behavior
+		q.JobIDConfig.JobID != "" {
+		return nil, fmt.Errorf("QueryConfig incompatible with fastPath")
+	}
+	pfalse := false
+	qRequest := &bq.QueryRequest{
+		Query:              q.QueryConfig.Q,
+		Location:           q.Location,
+		UseLegacySql:       &pfalse,
+		MaximumBytesBilled: q.QueryConfig.MaxBytesBilled,
+		RequestId:          uid.NewSpace("request", nil).New(),
+		Labels:             q.Labels,
+	}
+	if q.QueryConfig.DisableQueryCache {
+		pfalse := false
+		qRequest.UseQueryCache = &pfalse
```

Comment:
Is there a reason why you need a new false pointer here as opposed to reusing pfalse from L394?

Reply:
Hah, good spot!  Corrected.